tokio/src/runtime/io/driver/uring.rs

```diff
@@ -2,33 +2,85 @@ use io_uring::{squeue::Entry, IoUring};
 use mio::unix::SourceFd;
 use slab::Slab;
 
+use crate::loom::sync::atomic::Ordering;
 use crate::runtime::driver::op::{Cancellable, Lifecycle};
 use crate::{io::Interest, loom::sync::Mutex};
 
 use super::{Handle, TOKEN_WAKEUP};
 
-use std::os::fd::AsRawFd;
+use std::os::fd::{AsRawFd, RawFd};
 use std::{io, mem, task::Waker};
 
 const DEFAULT_RING_SIZE: u32 = 256;
 
+#[repr(usize)]
+#[derive(Debug, PartialEq, Eq)]
+enum State {
+    Uninitialized = 0,
+    Initialized = 1,
+    Unsupported = 2,
+}
+
+impl State {
+    fn as_usize(self) -> usize {
+        self as usize
+    }
+
+    fn from_usize(value: usize) -> Self {
+        match value {
+            0 => State::Uninitialized,
+            1 => State::Initialized,
+            2 => State::Unsupported,
+            _ => unreachable!("invalid Uring state: {}", value),
+        }
+    }
+}
+
 pub(crate) struct UringContext {
-    pub(crate) uring: io_uring::IoUring,
+    pub(crate) uring: Option<io_uring::IoUring>,
     pub(crate) ops: slab::Slab<Lifecycle>,
 }
 
 impl UringContext {
     pub(crate) fn new() -> Self {
         Self {
             ops: Slab::new(),
-            // TODO: make configurable
-            uring: IoUring::new(DEFAULT_RING_SIZE).unwrap(),
+            uring: None,
         }
     }
 
+    pub(crate) fn ring(&self) -> &io_uring::IoUring {
+        self.uring.as_ref().expect("io_uring not initialized")
+    }
+
+    pub(crate) fn ring_mut(&mut self) -> &mut io_uring::IoUring {
+        self.uring.as_mut().expect("io_uring not initialized")
+    }
+
+    /// Perform `io_uring_setup` system call, and Returns true if this
+    /// actually initialized the io_uring.
+    ///
+    /// If the machine doesn't support io_uring, then this will return an
+    /// `ENOSYS` error.
+    pub(crate) fn try_init(&mut self) -> io::Result<bool> {
+        if self.uring.is_some() {
+            // Already initialized.
+            return Ok(false);
+        }
+
+        self.uring.replace(IoUring::new(DEFAULT_RING_SIZE)?);
```

Comment:
nit: Shall we assert the error kind is `std::io::ErrorKind::Unsupported`?

Reply:
According to [this](https://github.com/rust-lang/rust/blob/100199c9aa50b0c47b37c9c86335d68b2a77b535/library/std/src/sys/pal/unix/mod.rs#L237), for example `EINVAL`, which could be returned by io_uring, is not mapped to `std::io::ErrorKind::Unsupported`. I think it's fine to return errors other than `ENOSYS` directly to the user.